controllers/volumereplicationgroup_controller.go

```diff
@@ -1029,74 +1146,313 @@ func (v *VRGInstance) createVR(vrNamespacedName types.NamespacedName, state volr
+
+	msg := "All PVCs of the VolumeReplicationGroup are not available"
+	setVRGErrorCondition(&v.instance.Status.Conditions, v.instance.Generation, msg)
+}
+
+func (v *VRGInstance) checkVRStatus(volRep *volrep.VolumeReplication) error {
 	switch {
-	case v.instance.Spec.ReplicationState == volrep.Primary:
+	case v.instance.Spec.ReplicationState == ramendrv1alpha1.Primary:
 		return v.validateVRStatusAsPrimary(volRep)
-	case v.instance.Spec.ReplicationState == volrep.Secondary:
+	case v.instance.Spec.ReplicationState == ramendrv1alpha1.Secondary:
 		return v.validateVRStatusAsSecondary(volRep)
-	case v.instance.Spec.ReplicationState == volrep.Resync:
-		return fmt.Errorf("resync is not supported for VRG")
 	default:
+		msg := "VolumeReplicationGroup state invalid"
+		v.updateProtectedPVCCondition(volRep.Name, PVCError, msg)
+
 		return fmt.Errorf("invalid Replication State %s for VolumeReplicationGroup (%s:%s)",
 			string(v.instance.Spec.ReplicationState), v.instance.Name, v.instance.Namespace)
 	}
 }
 
-// So, VolumeReplication controller reflects the status of a VolRep resource as primary
-// only after the promotion of the associated storage volume to primary is successful.
-// Hence, checking whether the primary state is reflected in the status is sufficient
-// to verify whether the VolRep resource is in a healthy state not. However, checking
-// conditions of VolRep status may be required for maintaining correct status of VRG
-// when VRG status is implemented.
-//
-// TODO: If checking conditions of VolRep is necessary for VRG status, add that logic
-//       here when VRG status is implemented.
+// VolumeReplication controller sets volRep.Status.State to primary only after the
+// successful promotion of the underlying storage volume consumed by the pvc.
+// It is necessary to get the appropriate condition out of different conditions
+// a VolRep status can have. Doing this may be helpful in getting a more
+// clear view of the VolumeReplication resource's current situation.
 func (v *VRGInstance) validateVRStatusAsPrimary(volRep *volrep.VolumeReplication) error {
+	// As of now, name of the VolumeReplication resource and pvc that it
+	// represents is same. So, using the name from volRep resource. In
+	// future if things change, some changes might have to be necessary.
 	if volRep.Status.State != volrep.PrimaryState {
 		v.log.Info("State is primary and status is not primary")
 
-		return fmt.Errorf("primary VolumeReplication resource status not same %s/%s (status: %s)",
-			volRep.Name, volRep.Namespace, volRep.Status.State)
+		// VolumeReplication status not being primary by this
+		// time means something is not correct. Generation has
+		// already been checked and without generation of the
+		// resource and its status matching, it would not have
+		// come here. So, mark it as error.
+		msg := "VolumeReplication status state and spec state not same"
+		v.updateProtectedPVCCondition(volRep.Name, PVCError, msg)
+
+		return nil
+	}
```

Comment:
@ShyamsundarR I am also thinking of removing the above check now that we are using conditions to determine whether the VoLRep is good or not. Whatever the information or result the above check provides, I expect that can be obtained even from checking VolRep conditions (done below) as well and simplifies the fact that VRG relies only on VOlRep conditions for maintaining its status. Do you agree?

Reply:
Ack!